File: services/gps-ingestor/src/routes/positions.js
'use strict';
const { Router } = require('express');
const { publishPosition } = require('../kafka/producer');
const { positionsTotal, positionsErrors } = require('../metrics');

const router = Router();

router.post('/', async (req, res) => {
  const { deliveryId, parcelId, latitude, longitude, speed, timestamp } = req.body;

  if (!parcelId || latitude == null || longitude == null) {
    return res.status(400).json({ error: 'parcelId, latitude and longitude are required' });
  }

  const position = {
    deliveryId: deliveryId || 'unknown',
    parcelId,
    latitude: parseFloat(latitude),
    longitude: parseFloat(longitude),
    speed: speed != null ? parseFloat(speed) : 0,
    timestamp: timestamp || new Date().toISOString()
  };

  try {
    await publishPosition(position);
    positionsTotal.inc();
    const eventId = `evt-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
    res.status(202).json({ accepted: true, eventId });
  } catch (err) {
    positionsErrors.inc();
    console.error('[gps-ingestor] Kafka publish error:', err.message);
    res.status(500).json({ error: 'Failed to publish position' });
  }
});

module.exports = router;

File: services/gps-ingestor/src/routes/positions.py
import os
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, HTTPException

from ..kafka.producer import producer
from ..metrics import POSITIONS_ERRORS, POSITIONS_TOTAL
from ..schemas import PositionInput

router = APIRouter()
TOPIC = os.environ.get("TOPIC_GPS_POSITIONS", "gps.positions")


@router.post("", status_code=202)
async def ingest_position(body: PositionInput):
    event_id = str(uuid.uuid4())
    event = {
        "eventId": event_id,
        "parcelId": body.parcelId,
        "latitude": body.latitude,
        "longitude": body.longitude,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }
    try:
        await producer.publish(TOPIC, event)
        POSITIONS_TOTAL.inc()
    except Exception as exc:
        POSITIONS_ERRORS.inc()
        raise HTTPException(status_code=503, detail="Message broker unavailable") from exc
    return {"eventId": event_id, "status": "queued"}
